File: backend/app/collectors/docker.py
from datetime import datetime

import docker
from docker.errors import DockerException

from app.collectors.base import BaseCollector
from app.config import DockerSystemConfig
from app.models.metrics import (
    ContainerMetrics,
    DockerMetrics,
    SystemMetrics,
    SystemStatus,
    SystemType,
)


class DockerCollector(BaseCollector):
    def __init__(self, system_id: str, name: str, config: DockerSystemConfig):
        super().__init__(system_id, name, SystemType.DOCKER)
        self.config = config
        self._client: docker.DockerClient | None = None

    def _get_client(self) -> docker.DockerClient:
        """Get or create Docker client."""
        if self._client is None:
            if self.config.host:
                self._client = docker.DockerClient(
                    base_url=self.config.host,
                    tls=self.config.tls,
                )
            else:
                self._client = docker.DockerClient(
                    base_url=self.config.socket,
                )
        return self._client

    async def check_connection(self) -> bool:
        """Check if Docker daemon is accessible."""
        try:
            client = self._get_client()
            client.ping()
            return True
        except DockerException:
            return False

    def _calculate_cpu_percent(self, stats: dict) -> float:
        """Calculate CPU percentage from container stats."""
        try:
            cpu_delta = (
                stats["cpu_stats"]["cpu_usage"]["total_usage"]
                - stats["precpu_stats"]["cpu_usage"]["total_usage"]
            )
            system_delta = (
                stats["cpu_stats"]["system_cpu_usage"]
                - stats["precpu_stats"]["system_cpu_usage"]
            )

            if system_delta > 0 and cpu_delta > 0:
                cpu_count = stats["cpu_stats"]["online_cpus"]
                return (cpu_delta / system_delta) * cpu_count * 100.0
        except (KeyError, ZeroDivisionError):
            pass
        return 0.0

    def _calculate_memory_percent(self, stats: dict) -> tuple[int, int, float]:
        """Calculate memory usage from container stats."""
        try:
            usage = stats["memory_stats"]["usage"]
            limit = stats["memory_stats"]["limit"]
            percent = (usage / limit) * 100.0 if limit > 0 else 0.0
            return usage, limit, percent
        except KeyError:
            return 0, 0, 0.0

    def determine_status(self, metrics: DockerMetrics) -> SystemStatus:
        """Determine system status based on Docker metrics."""
        if metrics.total_count == 0:
            return SystemStatus.HEALTHY

        unhealthy_count = 0
        high_resource_count = 0

        for container in metrics.containers:
            if container.state not in ("running", "paused"):
                continue
            if container.status == "unhealthy":
                unhealthy_count += 1
            if container.cpu_percent > 90 or container.memory_percent > 90:
                high_resource_count += 1

        if unhealthy_count > 0 or metrics.stopped_count > metrics.running_count:
            return SystemStatus.WARNING
        if high_resource_count > 0:
            return SystemStatus.WARNING

        return SystemStatus.HEALTHY

    async def collect(self) -> SystemMetrics:
        """Collect Docker container metrics."""
        client = self._get_client()

        containers = client.containers.list(all=True)
        container_metrics = []
        running_count = 0
        stopped_count = 0

        for container in containers:
            state = container.status

            if state == "running":
                running_count += 1
                try:
                    stats = container.stats(stream=False)
                    cpu_percent = self._calculate_cpu_percent(stats)
                    mem_usage, mem_limit, mem_percent = self._calculate_memory_percent(
                        stats
                    )
                except Exception:
                    cpu_percent = 0.0
                    mem_usage = mem_limit = 0
                    mem_percent = 0.0
            else:
                stopped_count += 1
                cpu_percent = 0.0
                mem_usage = mem_limit = 0
                mem_percent = 0.0

            health_status = "none"
            if container.attrs.get("State", {}).get("Health"):
                health_status = container.attrs["State"]["Health"].get(
                    "Status", "none"
                )

            container_metrics.append(
                ContainerMetrics(
                    id=container.short_id,
                    name=container.name,
                    image=container.image.tags[0] if container.image.tags else "unknown",
                    status=health_status,
                    state=state,
                    cpu_percent=round(cpu_percent, 2),
                    memory_usage=mem_usage,
                    memory_limit=mem_limit,
                    memory_percent=round(mem_percent, 2),
                )
            )

        docker_metrics = DockerMetrics(
            containers=container_metrics,
            running_count=running_count,
            stopped_count=stopped_count,
            total_count=len(containers),
        )

        return SystemMetrics(
            id=self.system_id,
            name=self.name,
            type=SystemType.DOCKER,
            status=self.determine_status(docker_metrics),
            last_updated=datetime.utcnow(),
            metrics=docker_metrics,
        )

File: backend/src/collectors/docker.rs
use async_trait::async_trait;
use bollard::container::{ListContainersOptions, StatsOptions};
use bollard::Docker;
use chrono::Utc;
use futures_util::TryStreamExt;

use super::Collector;
use crate::config::DockerSystemConfig;
use crate::models::*;

pub struct DockerCollector {
    system_id: String,
    name: String,
    config: DockerSystemConfig,
    client: Option<Docker>,
}

impl DockerCollector {
    pub fn new(system_id: String, name: String, config: DockerSystemConfig) -> Self {
        Self {
            system_id,
            name,
            config,
            client: None,
        }
    }

    fn get_client(&mut self) -> Result<&Docker, String> {
        if self.client.is_none() {
            let client = if let Some(ref host) = self.config.host {
                Docker::connect_with_http(host, 10, bollard::API_DEFAULT_VERSION)
                    .map_err(|e| format!("Docker connect failed: {e}"))?
            } else {
                Docker::connect_with_socket(
                    &self.config.socket,
                    10,
                    bollard::API_DEFAULT_VERSION,
                )
                .map_err(|e| format!("Docker socket connect failed: {e}"))?
            };
            self.client = Some(client);
        }
        Ok(self.client.as_ref().unwrap())
    }

    fn calculate_cpu_percent(stats: &bollard::container::Stats) -> f64 {
        let cpu_delta = stats.cpu_stats.cpu_usage.total_usage as f64
            - stats.precpu_stats.cpu_usage.total_usage as f64;
        let system_delta = stats.cpu_stats.system_cpu_usage.unwrap_or(0) as f64
            - stats.precpu_stats.system_cpu_usage.unwrap_or(0) as f64;

        if system_delta > 0.0 && cpu_delta > 0.0 {
            let cpu_count = stats.cpu_stats.online_cpus.unwrap_or(1) as f64;
            (cpu_delta / system_delta) * cpu_count * 100.0
        } else {
            0.0
        }
    }

    fn calculate_memory(stats: &bollard::container::Stats) -> (i64, i64, f64) {
        let usage = stats.memory_stats.usage.unwrap_or(0) as i64;
        let limit = stats.memory_stats.limit.unwrap_or(0) as i64;
        let percent = if limit > 0 {
            (usage as f64 / limit as f64) * 100.0
        } else {
            0.0
        };
        (usage, limit, percent)
    }

    fn determine_status(metrics: &DockerMetrics) -> SystemStatus {
        if metrics.total_count == 0 {
            return SystemStatus::Healthy;
        }

        let mut unhealthy_count = 0;
        let mut high_resource_count = 0;

        for container in &metrics.containers {
            if container.state != "running" && container.state != "paused" {
                continue;
            }
            if container.status == "unhealthy" {
                unhealthy_count += 1;
            }
            if container.cpu_percent > 90.0 || container.memory_percent > 90.0 {
                high_resource_count += 1;
            }
        }

        if unhealthy_count > 0 || metrics.stopped_count > metrics.running_count {
            return SystemStatus::Warning;
        }
        if high_resource_count > 0 {
            return SystemStatus::Warning;
        }

        SystemStatus::Healthy
    }

    fn create_error_metrics(&self, error: &str) -> SystemMetrics {
        SystemMetrics {
            id: self.system_id.clone(),
            name: self.name.clone(),
            system_type: SystemType::Docker,
            status: SystemStatus::Offline,
            last_updated: Utc::now(),
            error: Some(error.to_string()),
            metrics: MetricsPayload::default(),
        }
    }
}

#[async_trait]
impl Collector for DockerCollector {
    async fn collect(&mut self) -> SystemMetrics {
        let client = match self.get_client() {
            Ok(c) => c.clone(),
            Err(e) => return self.create_error_metrics(&e),
        };

        let containers = match client
            .list_containers(Some(ListContainersOptions::<String> {
                all: true,
                ..Default::default()
            }))
            .await
        {
            Ok(c) => c,
            Err(e) => return self.create_error_metrics(&e.to_string()),
        };

        let mut container_metrics = Vec::new();
        let mut running_count = 0i32;
        let mut stopped_count = 0i32;

        for container in &containers {
            let state = container.state.as_deref().unwrap_or("unknown").to_string();
            let name = container
                .names
                .as_ref()
                .and_then(|n| n.first())
                .map(|n| n.trim_start_matches('/').to_string())
                .unwrap_or_else(|| "unknown".to_string());
            let id = container
                .id
                .as_ref()
                .map(|id| id.chars().take(12).collect())
                .unwrap_or_else(|| "unknown".to_string());
            let image = container
                .image
                .as_deref()
                .unwrap_or("unknown")
                .to_string();

            let (cpu_percent, mem_usage, mem_limit, mem_percent) = if state == "running" {
                running_count += 1;
                if let Some(ref cid) = container.id {
                    match client
                        .stats(
                            cid,
                            Some(StatsOptions {
                                stream: false,
                                one_shot: true,
                            }),
                        )
                        .try_next()
                        .await
                    {
                        Ok(Some(stats)) => {
                            let cpu = Self::calculate_cpu_percent(&stats);
                            let (usage, limit, pct) = Self::calculate_memory(&stats);
                            (cpu, usage, limit, pct)
                        }
                        _ => (0.0, 0, 0, 0.0),
                    }
                } else {
                    (0.0, 0, 0, 0.0)
                }
            } else {
                stopped_count += 1;
                (0.0, 0, 0, 0.0)
            };

            let health_status = container
                .status
                .as_deref()
                .map(|s| {
                    if s.contains("healthy") && !s.contains("unhealthy") {
                        "healthy"
                    } else if s.contains("unhealthy") {
                        "unhealthy"
                    } else {
                        "none"
                    }
                })
                .unwrap_or("none")
                .to_string();

            container_metrics.push(ContainerMetrics {
                id,
                name,
                image,
                status: health_status,
                state,
                cpu_percent: (cpu_percent * 100.0).round() / 100.0,
                memory_usage: mem_usage,
                memory_limit: mem_limit,
                memory_percent: (mem_percent * 100.0).round() / 100.0,
            });
        }

        let docker_metrics = DockerMetrics {
            containers: container_metrics,
            running_count,
            stopped_count,
            total_count: containers.len() as i32,
        };

        let status = Self::determine_status(&docker_metrics);

        SystemMetrics {
            id: self.system_id.clone(),
            name: self.name.clone(),
            system_type: SystemType::Docker,
            status,
            last_updated: Utc::now(),
            error: None,
            metrics: MetricsPayload::Docker(docker_metrics),
        }
    }

    async fn close(&mut self) {
        self.client = None;
    }
}
